### PR TITLE
Add `{From,To}HttpApiData` instances for `Iso8601Range`.

### PR DESCRIPTION
**Pull request by jonathanknowles**
# Issue Number

#466 

# Overview

I have:

- [x] Added `{From,To}Text` instances for `Iso8601Range`.
- [x] Added roundtrip tests for `{From,To}Text` instances for `Iso8601Range`.
- [x] Added set of negative tests for `{From,To}Text` instances for `Iso8601Range`.
- [x] Added `{From,To}HttpApiData` instances for `Iso8601Range`.
- [x] Added roundtrip tests for `HttpApiData` instances for `Iso8601Range`.

### CODE REVIEW

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs [hunk 1]
```diff
@@ -546,6 +552,15 @@ instance Arbitrary ApiFee where
 instance Arbitrary AddressPoolGap where
     arbitrary = arbitraryBoundedEnum
 
+instance Arbitrary (Iso8601Range (name :: Symbol)) where
+    arbitrary = do
+        (mt1, mt2) <- arbitrary
+        let (mv1, mv2) =
+                if all isJust [mt1, mt2]
+                then (min mt1 mt2, max mt1 mt2)
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs]: Not sure to get this one? This seems to be forbidding `mt1` to be greater than `mt2` :thinking: ? (whereas that's a valid case for the range..)

File: lib/core/src/Cardano/Wallet/Api/Types.hs [hunk 2]
```diff
@@ -239,13 +248,65 @@ data ApiErrorCode
...
+            "Error encountered while decoding ISO 8601 time range: " <> err
+
+        parseTime :: Text -> Maybe (Maybe UTCTime)
+        parseTime = \case
+            "*" -> pure Nothing
+            timeText -> pure <$>
+                parseTimeM False defaultTimeLocale basicUtc $ T.unpack timeText
+
+        validRange :: Maybe UTCTime -> Maybe UTCTime -> Bool
+        validRange mt1 mt2 = fromMaybe True $ (<) <$> mt1 <*> mt2
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Types.hs]: Ah! Found it! 
This is _false_, the order between `mt1` and `mt2` conveys the ordering (ASC vs DESC) of the range, cf:

- `20190227T000000Z-20200227T000000Z`: means all transaction between 2019-02-27 and 2020-02-27, in ascending order.

- `20200227T000000Z-20190227T000000Z`: means all transaction between 2020-02-27 and 2019-02-27, in descending order.

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs (hunk 1 above)
Comment (jonathanknowles) [reply lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs]: > Not sure to get this one? This seems to be forbidding `mt1` to be greater than `mt2` thinking ? (whereas that's a valid case for the range..)

Hmm. I don't see how that's a valid case.

From my understanding of `Iso8601Range v1 v2`:

* `v1` is an optional _earliest time_.
* `v2` is an optional _latest time_.

So this `Arbitrary` instance is designed to only generate the following cases:

* `Iso8601Range Nothing Nothing`
* `Iso8601Range Nothing (Just t)`
* `Iso8601Range (Just t) Nothing`
* `Iso8601Range (Just t1) (Just t2)` where `(t1 <= t2)`.

Perhaps I have misunderstood?

File: lib/core/src/Cardano/Wallet/Api/Types.hs (hunk 2 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Api/Types.hs]: > Ah! Found it!
> This is _false_, the order between `mt1` and `mt2` conveys the ordering (ASC vs DESC) of the range, cf:
> 
>     * `20190227T000000Z-20200227T000000Z`: means all transaction between 2019-02-27 and 2020-02-27, in ascending order.
> 
>     * `20200227T000000Z-20190227T000000Z`: means all transaction between 2020-02-27 and 2019-02-27, in descending order.

Okay, that makes sense. I'll update the code to allow both ascending and descending ranges.

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs (hunk 1 above)
Comment (jonathanknowles) [reply lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs]: > Not sure to get this one? This seems to be forbidding `mt1` to be greater than `mt2` thinking ? (whereas that's a valid case for the range..)

I've just read your later comment. I'll update the code to support both ascending and descending ranges.

File: lib/core/src/Cardano/Wallet/Api/Types.hs (hunk 2 above)
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Api/Types.hs]: @KtorZ Now updated to allow ranges in both ascending and descending order.

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs (hunk 1 above)
Comment (jonathanknowles) [reply lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs]: @KtorZ Now updated to allow ranges in both ascending and descending order.

**Comment by paweljakubas** [comment]
@jonathanknowles is this ready for final review ? :-)
some tests still missing, right?

**Comment by jonathanknowles** [comment]
> @jonathanknowles is this ready for final review ? :-)
> some tests still missing, right?

Nearly ready! Will let you know when finished.

File: lib/core/src/Cardano/Wallet/Api/Types.hs [hunk 3]
```diff
@@ -220,32 +227,121 @@ data ApiErrorCode
...
+-- == Fully-open time ranges
+--
+-- * Queries with the following range will return data in __ascending__ order
+--   from the __start of time__ until the __end of time__:
+--
+--     > Iso8601Range Nothing Nothing
+--
+--     Example textual representation:
+--
+--     > "name *-*"
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Api/Types.hs:292]: very nice instructions describing how to use it :100:

File: lib/core/src/Cardano/Wallet/Api/Types.hs [hunk 4]
```diff
@@ -220,32 +227,121 @@ data ApiErrorCode
...
+            "Invalid prefix string found. Expecting: " <> show expectedPrefix
+        v1 <- guardM (parseTime t1) $
+            "Invalid start time string: " <> show t1
+        v2 <- guardM (parseTime t2) $
+            "Invalid end time string: " <> show t2
+        pure $ Iso8601Range v1 v2
+      where
+        expectedPrefix = T.pack $ symbolVal (Proxy @name)
+
+        guardB :: Bool -> String -> Either TextDecodingError ()
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Api/Types.hs:323]: probably there are already functions for that

File: lib/core/src/Cardano/Wallet/Api/Types.hs [hunk 5]
```diff
@@ -220,32 +227,121 @@ data ApiErrorCode
...
+        guardB b err = if b then Right () else raise err
+
+        guardM :: Maybe a -> String -> Either TextDecodingError a
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Api/Types.hs:326]: looks like : `https://hackage.haskell.org/package/either-5.0.1.1/docs/Data-Either-Combinators.html#v:maybeToRight`

**Review by paweljakubas — APPROVED** [review]
Cool! Small remarks to maybe use `maybeToEither`

**Pull request merged**